File: ScorpioConversion/src/Reader/DefaultReader.cs
using System;
using System.Text;
using System.IO;

public class DefaultReader : IReader, IDisposable {
    private readonly static DateTime BaseTime = new DateTime(1970, 1, 1, 0, 0, 0, 0, DateTimeKind.Utc);
    MemoryStream stream;
    BinaryReader reader;
    public string Name => "default";
    public DefaultReader() {
        
    }
    public void Initialize(byte[] buffer) {
        stream = new MemoryStream(buffer);
        reader = new BinaryReader(stream);
    }
    public void ReadHead(string fileName) {
        {
            var number = ReadInt32();        //字段数量
            for (var i = 0; i < number; ++i) {
                if (ReadInt8() == 0) {   //基础类型
                    ReadInt8();          //基础类型索引
                } else {                        //自定义类
                    ReadString();        //自定义类名称
                }
                ReadBool();          //是否是数组
            }
        }
        {
            var customNumber = ReadInt32();  //自定义类数量
            for (var i = 0; i < customNumber; ++i) {
                ReadString();                //读取自定义类名字
                var number = ReadInt32();        //字段数量
                for (var j = 0; j < number; ++j) {
                    if (ReadInt8() == 0) {   //基础类型
                        ReadInt8();          //基础类型索引
                    } else {                        //自定义类
                        ReadString();        //自定义类名称
                    }
                    ReadBool();          //是否是数组
                }
            }
        }
    }
    public bool ReadBool() {
        return ReadInt8() == 1;
    }
    public sbyte ReadInt8() {
        return reader.ReadSByte();
    }
    public byte ReadUInt8() {
        return reader.ReadByte();
    }
    public short ReadInt16() {
        return reader.ReadInt16();
    }
    public ushort ReadUInt16() {
        return reader.ReadUInt16();
    }
    public int ReadInt32() {
        return reader.ReadInt32();
    }
    public uint ReadUInt32() {
        return reader.ReadUInt32();
    }
    public long ReadInt64() {
        return reader.ReadInt64();
    }
    public ulong ReadUInt64() {
        return reader.ReadUInt64();
    }
    public float ReadFloat() {
        return reader.ReadSingle();
    }
    public double ReadDouble() {
        return reader.ReadDouble();
    }
    public string ReadString() {
        var length = reader.ReadUInt16();
        if (length <= 0) return "";
        return Encoding.UTF8.GetString(reader.ReadBytes(length));
    }
    public DateTime ReadDateTime() {
        DateTime startTime = BaseTime;
        return startTime.AddMilliseconds(reader.ReadInt64());
    }
    public byte[] ReadBytes() {
        var length = reader.ReadInt32();
        return reader.ReadBytes(length);
    }
    public void Dispose() {
        stream.Close();
        reader.Close();
    }
}

File: ScorpioProto/Python/Scorpio.Conversion.Runtime/ScorpioConversionRuntime/DefaultReader.py
from struct import pack, unpack
import time
class DefaultReader:
    def __init__(self, base_stream):
        self.base_stream = base_stream
        
    def readBytes(self, length):
        return self.base_stream.read(length)

    def unpack(self, fmt, length = 1):
        return unpack(fmt, self.readBytes(length))[0]

    def ReadBool(self):
        return self.ReadInt8() == 1

    def ReadInt8(self):
        return self.unpack('b')

    def ReadUInt8(self):
        return self.unpack('B')

    def ReadInt16(self):
        return self.unpack('h', 2)

    def ReadUInt16(self):
        return self.unpack('H', 2)

    def ReadInt32(self):
        return self.unpack('i', 4)

    def ReadUInt32(self):
        return self.unpack('I', 4)

    def ReadInt64(self):
        return self.unpack('q', 8)

    def ReadUInt64(self):
        return self.unpack('Q', 8)

    def ReadFloat(self):
        return self.unpack('f', 4)

    def ReadDouble(self):
        return self.unpack('d', 8)

    def ReadString(self):
        length = self.ReadUInt16()
        return self.readBytes(length).decode("utf-8")

    def ReadDateTime(self):
        # return datetime.fromtimestamp(int(self.ReadInt64()))
        # return self.ReadInt64()
        return time.localtime(self.ReadInt64() / 1000)
        
    def ReadBytes(self):
        length = self.ReadInt32()
        return self.readBytes(length)

    def ReadHead(this):
        number = this.ReadInt32()           #字段数量
        for i in range(0, number):
            if this.ReadInt8() == 0:        #基础类型
                this.ReadInt8()             #基础类型索引
            else:                           #自定义类
                this.ReadString()           #自定义类名称
            this.ReadBool()                 #是否是数组
            this.ReadString()               #字段名称

        customNumber = this.ReadInt32()             #自定义类数量
        for i in range(0, customNumber):
            this.ReadString()                       #读取自定义类名字
            if this.ReadInt8() == 1:
                number = this.ReadInt32()
                for j in range(0, number):
                    this.ReadString()
                    this.ReadInt32()
            else:
                number = this.ReadInt32()           #字段数量
                for j in range(0, number):
                    if this.ReadInt8() == 0:        #基础类型
                        this.ReadInt8()             #基础类型索引
                    else:                           #自定义类
                        this.ReadString()           #自定义类名称
                    this.ReadBool()                 #是否是数组
                    this.ReadString()

    def Close(self):
        pass